### DISCUSSION TITLE
[BUG] Hotkeys Don't Work in Some Games

### DISCUSSION THREAD

**Discussion by Switch123456789** [radj307/volume-control#49] [Support]
### Version
5.2.2

### Actual Behaviour
hotkeys don't work while the game "Euro Truck Simulator 2" (DX11 x64) is focused (Windows 11).

### Reproduction Steps
1. focus ETS2
2. try hotkeys (they do not work)

**Comment by radj307** [comment]
This is a problem in some games due to the methods they use for capturing input; however this can be changed *(or broken, depending on how you look at it)* by external factors, such as your window mode.  
If you're using **Fullscreen**, try **Borderless Windowed**, or vica-versa.

Other external factors are injected, such as reshade, since those often have to interfere with the games own handling of input.  For example, hotkeys don't work in Skyrim Special Edition either, but with the recent release of Improved Camera 4 *(which implements its own IMGUI-style UI injection)*, that actually *fixes* the problem.

I would try changing the window mode first, and if that doesn't work *(or if it isn't supported)* try using reshade since its more lightweight than the forced-borderless apps out there.

Unfortunately I don't know of any others ways to circumvent this without implementing it at the driver level, which is something I have absolutely no experience with.

**Comment by Switch123456789** [comment]
Hey, I appreciate the quick response and infos :)
I just tried it in windowed mode but that didn't work either unfortunately. I will try reshade soon and report back.

Previously I was using [ETS2/ATS Local Radio](https://github.com/Koenvh1/ets2-local-radio), it can be controlled within ETS2 using hotkeys (without modifying the game itself), maybe you can find something within the code that will give you a hint how to fix it for those kind of games. I am not an experienced programmer by any means so please forgive me if this idea is/sounds bonkers.

**Comment by radj307** [comment]
Thanks for the link - that is actually really helpful!  

That mod works by injecting a DLL into the ETS game process at runtime in order to intercept inputs made via the **DirectInput API**.  

In my opinion this is the single most annoying drawback of Volume Control, so it's certainly a priority - but it'll likely end up being a longer-term development milestone rather than something that can be fixed right away due to all of the development required to get this to reliably work for a wide range of games.  

Perhaps this could be solved using an external solution, such as a generic injected input-pipe that sends key presses made via dinput to the OS if certain conditions are met? Just brainstorming on that, but who knows, maybe it already exists somewhere.

**Comment by Switch123456789** [comment]
> maybe it already exists somewhere.

This made me searching for "it" and I what I found was a bit shocking. Apparently running volume-control in admin-mode is all it takes to work around that issue. At least for ETS2 it is.
